md/groovy-in-bonita.md

```diff
@@ -0,0 +1,347 @@
+    String name
+}
+ 
+class Address {
+    Street street
+    String postalCode
+    String city
+}
+ 
+class Street {
+    String name
+    String number
+    String additionalInfo
+}
+```
+
+We want to display the street name, but we don't know if all object instances are available. To avoid a NullPointerException we write the following code:
+
+```groovy
+// company can be null.
+if (company != null && company.getAddress() != null && company.getAddress().getStreet() != null) {
+    println company.address.street.name
+}
+```
+
+Groovy adds the safe navigation operator to shorten all this to:
+
+```groovy
+// company can be null.
+println company?.address?.street?.name
+```
+
+* The Elvis operator `:?` to shorten ternary expression
+
+```groovy
+def sampleText
+ 
+// Normal ternary operator.
+def ternaryOutput = (sampleText != null) ? sampleText : 'Hello Groovy!'
+ 
+// The Elvis operator in action. We must read: 'If sampleText is not null assign
+// sampleText to elvisOuput, otherwise assign 'Viva Las Vegas!' to elvisOutput.
+def elvisOutput = sampleText ?: 'Viva Las Vegas!'
+```
+
+More about Operators in Groovy [here](http://docs.groovy-lang.org/docs/groovy-2.4.16/html/documentation/#groovy-operators).
+
+## Bonita use cases with Groovy
+
+In the below examples, the following BDM will be used
+
+`*` for mandatory fields
+
+```groovy
+Comment {
+    String content*
+    AppUser createdBy* //Aggregation
+    DateTime creationDate*
+    AppUser lastEditedBy* //Aggregation
+}
+
+AppUser {
+    String firstName*
+    String lastName*
+    Address address //Composition
+}
+
+Address {
+    Street street* //Composition
+    String postalCode*
+    String city*
+}
+ 
+Street {
+    String name*
+    String number*
+    String additionalInfo
+}
+```
+
+### Instantiate a Business Data
+
+#### Using the generated DAO
+
+By default a `newInstance` factory method is generated in the object DAO.  
+This method has as many parameters as mandatory fields for this object.
+
+```groovy
+appUserDAO.newInstance('Jane','Doe') // create a UserApp instance with firstName = 'Jane' and lastName = 'Doe'
+```
+
+#### Using a constructor with named arguments
+
+```groovy
+new Address(street: new Street(number: '32', name: 'Gustave Eiffel'), postalCode: '38000', city: 'Grenoble')
+```
+
+#### From a map
+
+Given a complex contract input
+
+```
+userInput COMPLEX
+	firstName TEXT
+	lastName TEXT
+```
+
+As a COMPLEX input is stored as Map it is possible to write
+
+```groovy
+userInput as UserApp
+```
+
+### List Business Objects using DAO
+
+When defining your BDM you can write custom queries in JPQL than can be called using the object DAO.
+Some queries are generated by default like: `find`, `findByFirstName`, `findByLastName`...
+
+All object DAO are injected at runtime in Groovy script expression or can be retrieved using [APIClient#getDAO](https://documentation.bonitasoft.com/javadoc/api/7.9/org/bonitasoft/engine/api/APIClient.html#getDAO-java.lang.Class-)
```

Comment:
Please use `${varVersion}` instead of `7.9`

Reply:
❤️